staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go

```diff
@@ -755,90 +755,98 @@ func TestStreamTranslator_WebSocketServerErrors(t *testing.T) {
-		t.Fatal(err)
+	if err := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 2*time.Second, true, func(ctx context.Context) (bool, error) {
+		if testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expected), metricNames...) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to observe metric after waiting 2 seconds: %v", err)
 	}
 }
 
 // TestStreamTranslator_BlockRedirects verifies that the StreamTranslator will *not* follow
 // redirects; it will thrown an error instead.
 func TestStreamTranslator_BlockRedirects(t *testing.T) {
 	metrics.Register()
-	metrics.ResetForTest()
 	t.Cleanup(metrics.ResetForTest)
+
 	for _, statusCode := range []int{
 		http.StatusMovedPermanently,  // 301
 		http.StatusFound,             // 302
 		http.StatusSeeOther,          // 303
 		http.StatusTemporaryRedirect, // 307
 		http.StatusPermanentRedirect, // 308
 	} {
-		// Create upstream fake SPDY server which returns a redirect.
-		spdyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			w.Header().Set("Location", "/")
-			w.WriteHeader(statusCode)
-		}))
-		defer spdyServer.Close()
-		spdyLocation, err := url.Parse(spdyServer.URL)
-		if err != nil {
-			t.Fatalf("Unable to parse spdy server URL: %s", spdyServer.URL)
-		}
-		spdyTransport, err := fakeTransport()
-		if err != nil {
-			t.Fatalf("Unexpected error creating transport: %v", err)
-		}
-		streams := Options{Stdout: true}
-		streamTranslator := NewStreamTranslatorHandler(spdyLocation, spdyTransport, 0, streams)
-		streamTranslatorServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			streamTranslator.ServeHTTP(w, req)
-		}))
-		defer streamTranslatorServer.Close()
-		// Now create the websocket client (executor), and point it to the "streamTranslatorServer".
-		streamTranslatorLocation, err := url.Parse(streamTranslatorServer.URL)
-		if err != nil {
-			t.Fatalf("Unable to parse StreamTranslator server URL: %s", streamTranslatorServer.URL)
-		}
-		exec, err := remotecommand.NewWebSocketExecutor(&rest.Config{Host: streamTranslatorLocation.Host}, "GET", streamTranslatorServer.URL)
-		if err != nil {
-			t.Errorf("unexpected error creating websocket executor: %v", err)
-		}
-		errorChan := make(chan error)
-		go func() {
-			// Start the streaming on the WebSocket "exec" client.
-			// Should return "redirect not allowed" error.
-			errorChan <- exec.StreamWithContext(context.Background(), remotecommand.StreamOptions{})
-		}()
-
-		select {
-		case <-time.After(wait.ForeverTestTimeout):
-			t.Fatalf("expect stream to be closed after connection is closed.")
-		case err := <-errorChan:
-			// Must return "redirect now allowed" error.
-			if err == nil {
-				t.Fatalf("expected error, but received none")
+		t.Run(fmt.Sprintf("statusCode=%d", statusCode), func(t *testing.T) {
+			metrics.ResetForTest()
```

Comment:
reset before *and* after the test?

```
metrics.ResetForTest()
t.Cleanup(metrics.ResetForTest)
```

Reply:
not really needed, the next subtest will reset them before start